### PR TITLE
add Script Id feature

### PR DESCRIPTION
**Pull request by nicolassing**
I've also refactorize all Scripting type

### CODE REVIEW

File: lib/Elastica/AbstractScript.php [hunk 1]
```diff
@@ -1,21 +1,36 @@
...
+    protected $script;
+
+    protected $lang;
+
     /**
      * @param array|null $params
      * @param string     $id
      */
-    public function __construct(array $params = null, $id = null)
+    public function __construct($script, array $params = null, $lang = null, $id = null)
```
Comment (ruflin) [inline lib/Elastica/AbstractScript.php]: If I understand this right, this is only a BC break for people that extended from AbstractScript. For all the inherited classes there is no difference that is not part of the library?

File: lib/Elastica/Client.php [hunk 2]
```diff
@@ -716,4 +716,27 @@ public function setLogger(LoggerInterface $logger)
...
+    /**
+     * @param Script $script
+     * @param $lang
+     * @param $id
+     *
+     * @return Response
+     */
+    public function addIndexedScript(Script $script, $lang, $id)
+    {
+        return $this->request(sprintf('_scripts/%s/%s', $lang, $id), Request::POST, $script->toArray());
```
Comment (ruflin) [inline lib/Elastica/Client.php]: Can you use string concatenation here? `'_script/' . $lang . '/' . $id)` ? The main reason is that it is used like this across the library.

File: lib/Elastica/Client.php [hunk 3]
```diff
@@ -716,4 +716,27 @@ public function setLogger(LoggerInterface $logger)
...
+
+    /**
+     * @param $lang
+     * @param $id
+     *
+     * @return Response
+     */
+    public function getIndexedScript($lang, $id)
+    {
+        return $this->request(sprintf('_scripts/%s/%s', $lang, $id), Request::GET);
```
Comment (ruflin) [inline lib/Elastica/Client.php]: See above.

File: test/lib/Elastica/Test/ScriptFileTest.php [hunk 4]
```diff
@@ -53,7 +53,7 @@ public function testConstructor()
         $expected = array(
             'script_file' => $value,
         );
-        $this->assertEquals($value, $scriptFile->getScriptFile());
```
Comment (ruflin) [inline test/lib/Elastica/Test/ScriptFileTest.php]: As getScriptFile is not released yet I suggest we don't call this change a BC change. See https://github.com/ruflin/Elastica/blob/master/CHANGELOG.md Or should we create a proxy method getScriptFile in the scriptFile object?

**Comment by ruflin** [comment]
The changes look good. See my comments above. Can you also update the CHANGELOG.md file? 

I like that it seems to simplify the script part once again :+1:

**Comment by ruflin** [comment]
@nicolassing Any updates here? Would love to get this into master soon.

**Comment by Tobion** [comment]
Solved by #1242

**Pull request closed without merging**